src/binary_annotator.cpp

```diff
@@ -0,0 +1,647 @@
+
+  // TODO(dbaileychess): sized prefixed value
+  const uint32_t root_table_offset = GetScalar<uint32_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(
+      offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0, root_table_offset,
+      std::string("offset to root table `") +
+          schema_->root_table()->name()->str() + "`"));
+
+  const uint32_t file_identifier_value =
+      GetScalar<uint32_t>(offset + sizeof(uint32_t));
+
+  if (file_identifier_value) {
+    // Check if the file identifier region has non-zero data, and assume its the
+    // file identifier. Otherwise, it will get filled in with padding later.
+    regions.emplace_back(MakeBinaryRegion(
+        offset + sizeof(uint32_t), 4 * sizeof(uint8_t), BinaryRegionType::Char,
+        4, 0, std::string("File Identifier")));
+  }
+
+  sections_.insert(std::make_pair(
+      offset, MakeBinarySection("", BinarySectionType::Header, regions)));
+  return root_table_offset;
+}
+
+void BinaryAnnotator::BuildVTable(uint64_t offset,
+                                  const reflection::Object *table) {
+  const uint64_t vtable_offset = offset;
+
+  // First see if we have used this vtable before, if so skip building it again.
+  auto it = vtables_.find(vtable_offset);
+  if (it != vtables_.end()) { return; }
+
+  std::vector<BinaryRegion> regions;
+
+  // Vtables start with the size of the vtable
+  const uint16_t vtable_size = GetScalar<uint16_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(offset, sizeof(uint16_t),
+                                        BinaryRegionType::Uint16, 0, 0,
+                                        std::string("size of this vtable")));
+  offset += sizeof(uint16_t);
+
+  // Then they have the size of the table they reference.
+  regions.emplace_back(
+      MakeBinaryRegion(offset, sizeof(uint16_t), BinaryRegionType::Uint16, 0, 0,
+                       std::string("size of referring table")));
+  offset += sizeof(uint16_t);
+
+  const uint64_t offset_start = offset;
+
+  // A mapping between field (and its id) to the relative offset (uin16_t) from
+  // the start of the table.
+  std::vector<std::tuple<const reflection::Field *, uint16_t>> fields;
+
+  // Loop over all the fields.
+  ForAllFields(table, /*reverse=*/false, [&](const reflection::Field *field) {
+    const uint64_t field_offset = offset_start + field->id() * sizeof(uint16_t);
+
+    if (field_offset >= vtable_offset + vtable_size) {
+      // This field_offset is too large for this vtable, so it must come from a
+      // newer schema than the binary was create with. Ignore.
+
+      // TODO(dbaileychess): We could show which fields are not set an their
+      // default values if we want. We just need a way to make it obvious that
+      // it isn't part of the buffer.
+      return;
+    }
+
+    const uint16_t value = GetScalar<uint16_t>(field_offset);
+
+    fields.push_back(std::make_pair(field, GetScalar<uint16_t>(field_offset)));
+    regions.push_back(MakeBinaryRegion(
```

Comment:
I know you're not doing all of the advanced features, but could you at least leave a TODO?

`<defaults to 0>` is perhaps unhelpful for table/string/vector/optional scalar fields.
We might also want to label required fields

Reply:
> I know you're not doing all of the advanced features, but could you at least leave a TODO?

Not sure what you mean, this is just the vtable offsets, I should handle them all here.

> <defaults to 0> is perhaps unhelpful for table/string/vector/optional scalar fields.

Done.

> We might also want to label required fields

Made this an error annotation.